fix enum attribute

kmuehlbauer opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

<!--
    This fixes an issue with enum attributes, where instead of the value, the dtype was returned.
    Closes #101 
-->

This fixes an issue with enum attributes, where instead of the value, the dtype was returned.

Closes #101

While doing this, I had to enhance the tests and according make/data file. Also, I had to fix some inconsistencies in that particular test file wrt numpy.


## Before you get started

- [x] [☝ Create an issue](https://github.com/NCAS-CMS/pyfive/issues) to discuss what you are going to do

## Checklist

- [x] This pull request has a descriptive title and labels
- [x] This pull request has a minimal description (most was discussed in the issue, but a two-liner description is still desirable)
- [x] Unit tests have been added (if codecov test fails)
- [x] Any changed dependencies have been added or removed correctly (if need be)
- [ ] If you are working on the documentation, please ensure the [current build](https://app.readthedocs.org/projects/pyfive/builds/) passes
- [ ] All tests pass

kmuehlbauer commented on tests/test_attr_datatypes.py:
```diff
@@ -79,10 +79,14 @@ def test_numeric_array_attr_datatypes():
-        assert hfile.attrs['vlen_str_array'][1] == b'World!'
 
-        assert hfile.attrs['vlen_str_array'].dtype == np.dtype('S6')
+def test_string_array_attr_datatypes():
+    with pyfive.File(ATTR_DATATYPES_HDF5_FILE) as hfile:
+        assert hfile.attrs['vlen_str_array'][0] == 'Hello'
+        assert hfile.attrs['vlen_str_array'][1] == 'World!'
+
+        assert hfile.attrs['vlen_str_array'].dtype == np.dtype('O')
+        assert hfile.attrs['vlen_str_array'].dtype.metadata == {'h5py_encoding': 'utf-8'}
```
This reads back now as strings. Because I had to fix this, I've taken the chance to move this to its own test function.

bnlawrence replied:
So, I think this is a different behaviour from h5py. 
 ```
with h5py.File(ATTR_DATATYPES_HDF5_FILE) as hfile:
    ...:     x=hfile.attrs['vlen_str_array'][0]
    ...:     print(x)
b'Hello'
```
so I'm thinking this is not what we want?

kmuehlbauer replied:
I'll double check this one.

kmuehlbauer replied:
And from looking at this, the old version doesn't seem to be variable length at all.

kmuehlbauer commented on tests/make_attr_datatypes_file.py:
```diff
@@ -55,20 +55,20 @@
 attrs['vlen_str_array'] = [b'Hello', b'World!']
 
 # variables length sequences
-val = np.empty((2, ), dtype=np.object)
```
This was complained about in the test suite, so I fixed this as suggested.

kmuehlbauer commented on pyfive/dataobjects.py:
```diff
@@ -250,6 +250,12 @@ def _parse_attribute_msg(self, buffer, offset):
         """ Retrieve an HDF5 attribute value from a buffer. """
+
+        # first handle ENUMERATION, we just extract the dtype
+        if isinstance(dtype, tuple):
```
Extracting this in that early stage here works just fine. This then runs the normal path and seems to fix the specific problem for enum attributes.

kmuehlbauer reviewed (commented):
I've left some comments for clarification.

kmuehlbauer commented on pyfive/dataobjects.py:
```diff
@@ -230,11 +231,17 @@ def _parse_attribute_msg(self, buffer, offset):
 
-        # Read the value(s)
-        value = self._attr_value(dtype, buffer, items, offset)
+        # detect Empty/NULL dataspace
+        if shape is None:
+            value = Empty(dtype=dtype)
```
After detecting NULL dataspace in `determine_data_shape`, we create Empty value here.

bnlawrence replied:
It'd be good to fix the docstring for empty, as we're now actually using it!

kmuehlbauer replied:
Done!

kmuehlbauer commented on pyfive/dataobjects.py:
```diff
@@ -722,6 +733,11 @@ def determine_data_shape(buf, offset):
         offset += DATASPACE_MSG_HEADER_V2_SIZE
     else:
         raise InvalidHDF5File('unknown dataspace message version')
+    # to detect Empty aka NULL dataspace we need to check for V2 DATASPACE MESSAGE
+    if header["version"] == 2:
+        # check for NULL dataspace and return early
+        if header["type"] == 2:
+            return None, None
 
```
That's the crucial part, see https://support.hdfgroup.org/documentation/hdf5/latest/_f_m_t2.html#subsubsec_fmt2_dataobject_hdr_msg_simple section V2 for details.

bnlawrence replied:
I'd be tempted to move the `if header['type]] == 2; return None, None` into the version elif ... but I'm also happy to leave it here.

kmuehlbauer reviewed (commented):
I've added a fix for #100 as it fits thematically in here. Please let me know, if you are K with this.

valeriupredoi commented:
woohoo @kmuehlbauer you're an absolute legend, was bashing me brains about #100 on Friday, many thanks! Would you mind if I added an integration test with a real (but small) file that replicates that issue? :beers:

valeriupredoi commented:
> Yes, please go ahead. 🚀
> 
> I'm currently working on ironing out some more issues with EnumType and other user types (eg. COMPOUND).

brill! Should I just PR it to your fork? I still need to get the correct (small) file, I was banking on Iris to chop the original file - which I did, but Iris has fixed that NULL (which is interesting to see it doing that)

kmuehlbauer commented:
@valeriupredoi Just push any relevant changes directly to my PR. No need to fiddle with PR's here.

valeriupredoi reviewed (approved):
this fixes #100 very nicely - I tested both locally (POSIX) and remotely (S3), very many thanks @kmuehlbauer 🍻

kmuehlbauer commented:
> OK @kmuehlbauer I've been trying to write a file with that NULL DATATSPACE for a wee while now - no success, netCDF4 is actually writing it but adds a point to the data, and if you try pass a Null/None, it complains - do you know how to create one? Got no idea how that file was created to have such a peculiar attribute/null value 🍻

I've added one attribute `empty_string` to `attr_datatypes.hdf5` (and the respective make-file). I'm not sure, when this is triggered. as this attribute is just a global attribute, no attribute of a dataset. Not sure, if that makes a difference. But yes, you could check if this breaks without the fix.

valeriupredoi commented on tests/test_buffer_issue.py:
```diff
@@ -0,0 +1,38 @@
+    """
+    print("File with issue da193a_25_6hr_t_pt_cordex__198807-198807.nc")
+    print("Variable m01s30i111")
+    _load_nc_file('m01s30i111')
```
@kmuehlbauer in lieu of an actual small file I've added a test case for an S3-hosted file - this works beautifully with this PR

kmuehlbauer replied:
Nice!

valeriupredoi replied:
if you ready with this PR, I reckon we can merge, bud. We should definitely prepare a new release with this and #105

bnlawrence reviewed (commented):
I'm slightly concerned that pyfive will now be doing different things with strings? (The bug fix is fine though!)

kmuehlbauer reviewed (commented):
@bnlawrence I've worked along your suggestions.

kmuehlbauer commented:
@bnlawrence According to our discussion in #108 I've reverted the changes of the ingested hdf5 file. Instead I only changed the according make file, taking now an argument destination file path.

For testing I introduced a fixture, which creates that file on demand and tests the necessary added features in new functions.

Is this a way forward? Slight changes to the make scripts and only some changes to the tests.

bnlawrence commented:
Great. Thanks. @valeriupredoi do your thing.

valeriupredoi commented:
Very many thanks @kmuehlbauer and @bnlawrence 🍺🍺
